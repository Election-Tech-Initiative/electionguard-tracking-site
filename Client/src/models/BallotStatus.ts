export class BallotStatus {
  public trackingId?: string;
  public status?: number;
  public location?: string;
  public approximateCastTime?: Date;
  public details?: string[];

  public matchedSegment?: string;
  public unmatchedSegment?: string;
}
