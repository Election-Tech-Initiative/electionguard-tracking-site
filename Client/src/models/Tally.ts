export class Tally {
  public selectionId?: string;
  public name?: string;
  public description?: string;
  public party?: string;
  public voteCount?: number;
}
