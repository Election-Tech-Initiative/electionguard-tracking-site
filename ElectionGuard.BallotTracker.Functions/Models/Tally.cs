namespace ElectionGuard.BallotTracker.Functions.Models
{
    public class Tally
    {
        public string SelectionId { get; set; }

        public string Name { get; set; }

        public string Description { get; set; }

        public string Party { get; set; }

        public int VoteCount { get; set; }
    }
}