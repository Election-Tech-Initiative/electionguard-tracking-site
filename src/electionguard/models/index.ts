export * from './election_object_base';
export * from './election';
export * from './tally';
