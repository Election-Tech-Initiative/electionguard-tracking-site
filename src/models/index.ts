export * from './election';
export * from './internationalizedText';
export * from './tally';
export * from './tracking';
