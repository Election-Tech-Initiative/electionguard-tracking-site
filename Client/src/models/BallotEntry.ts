import { Tally } from './Tally';

export class BallotEntry {
  public raceId?: string;
  public name?: string;
  public description?: string;
  public tallies?: Tally[];
}
