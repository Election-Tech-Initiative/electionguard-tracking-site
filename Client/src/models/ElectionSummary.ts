import { BallotEntry } from './BallotEntry';

export class ElectionSummary {
  public isComplete!: boolean;
  public ballotEntries?: BallotEntry[];
}
