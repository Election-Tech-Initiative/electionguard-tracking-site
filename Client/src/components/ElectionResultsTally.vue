<template>
  <div id="tally">
    <h2 class="title">
      {{ $t('electionResults') }}
    </h2>
    <div v-for="(item) in electionResults.ballotEntries" :key="item.raceId">
      <hr class="hr" />
      <h3 class="title">{{item.name}}</h3>
      <div>
        <div class="columns"
          v-for="(tally, index) in item.tallies"
          :key="tally.selectionId">
          <div class="name column is-one-quarter">
            {{tally.name}}
            <div class="party">{{tally.party}}</div>
          </div>
          <div class="votes column">
            <ElectionResultsTallyBar
              :isFirst="index === 0"
              :count="tally.voteCount"
              :percentage="getPercentage(tally.voteCount, item.tallies)"
            />
          </div>
        </div>
      </div>
    </div>
    <button class="button is-success tooltip" :data-tooltip="$t('featureNotAvailable')">{{ $t('downloadElectionTally') }}</button>    
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import ElectionResultsTallyBar from '@/components/ElectionResultsTallyBar.vue';
import { ElectionSummary } from '@/models/ElectionSummary';
import { Tally } from '@/models/Tally';

@Component({
  components: {
    ElectionResultsTallyBar,
  }
})
export default class ElectionResultsTally extends Vue {
  @Prop() public electionResults!: ElectionSummary;

  public getPercentage(count: number, tallies: Tally[]): string {
    const totalVotes = tallies
      .map((t) => t.voteCount)
      .reduce((total, votes) => (total || 0) + (votes || 0));
    if (totalVotes) {
      const percentage = (count / totalVotes) * 100;
      return `${percentage.toFixed(1)}%`;
    }
    return '0%';
  }
}
</script>

<style lang="scss" scoped>
.button.is-success {
  margin-top: 5rem;
}

.columns {
  
  .name {
    padding-left: 4rem;

    @include touch {
      padding-left: 0.75rem;
    }
  }

  .party {
    color: $grey-dark;
    font-size: 0.8rem;
    font-weight: normal;
  }
}
</style>
