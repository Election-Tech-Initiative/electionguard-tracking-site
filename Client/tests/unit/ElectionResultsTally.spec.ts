import { mount } from '@vue/test-utils';

import ElectionResultsTally from '@/components/ElectionResultsTally.vue';
import { Tally } from '@/models/Tally';

jest.mock('@/components/ElectionResultsTallyBar.vue', () => ({ default: jest.fn() }));

describe('ElectionResultsTally.vue', () => {
    const getMountedWrapper = (tallies: Tally[]) => mount(ElectionResultsTally, {
        mocks: {
            $t: (msg: string): string => msg
        },
        propsData: {
            electionResults: {
                isComplete: true,
                ballotEntries: [
                    {
                        raceId: 'mockRaceId',
                        name: 'mockRaceName',
                        description: 'mockRaceDescription',
                        tallies,
                    },
                ],
            },
        },
        stubs: [
            'ElectionResultsTallyBar',
        ],
    });

    it('renders 25% and 75%', () => {
        const wrapper = getMountedWrapper([
            {
                selectionId: 'mockSelectionId0',
                name: 'mockName0',
                description: 'mockDescription0',
                party: 'mockParty0',
                voteCount: 1,
            },
            {
                selectionId: 'mockSelectionId1',
                name: 'mockName1',
                description: 'mockDescription1',
                party: 'mockParty1',
                voteCount: 3,
            },
        ]);
        expect(wrapper.element).toMatchSnapshot('with 25% and 75%');
    });

    it('renders 10% and 90%', () => {
        const wrapper = getMountedWrapper([
            {
                selectionId: 'mockSelectionId0',
                name: 'mockName0',
                description: 'mockDescription0',
                party: 'mockParty0',
                voteCount: 1,
            },
            {
                selectionId: 'mockSelectionId1',
                name: 'mockName1',
                description: 'mockDescription1',
                party: 'mockParty1',
                voteCount: 9,
            },
        ]);
        expect(wrapper.element).toMatchSnapshot('with 10% and 90%');
    });
});
