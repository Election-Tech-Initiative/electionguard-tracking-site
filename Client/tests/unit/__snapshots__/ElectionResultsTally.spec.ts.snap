// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectionResultsTally.vue renders 10% and 90%: with 10% and 90% 1`] = `
<div
  id="tally"
>
  <h2
    class="title"
  >
    
    electionResults
  
  </h2>
   
  <div>
    <hr
      class="hr"
    />
     
    <h3
      class="title"
    >
      mockRaceName
    </h3>
     
    <div>
      <div
        class="columns"
      >
        <div
          class="name column is-one-quarter"
        >
          
          mockName0
          
          <div
            class="party"
          >
            mockParty0
          </div>
        </div>
         
        <div
          class="votes column"
        >
          <electionresultstallybar-stub
            count="1"
            isfirst="true"
            percentage="10.0%"
          />
        </div>
      </div>
      <div
        class="columns"
      >
        <div
          class="name column is-one-quarter"
        >
          
          mockName1
          
          <div
            class="party"
          >
            mockParty1
          </div>
        </div>
         
        <div
          class="votes column"
        >
          <electionresultstallybar-stub
            count="9"
            percentage="90.0%"
          />
        </div>
      </div>
    </div>
  </div>
   
  <button
    class="button is-success tooltip"
    data-tooltip="featureNotAvailable"
  >
    downloadElectionTally
  </button>
</div>
`;

exports[`ElectionResultsTally.vue renders 25% and 75%: with 25% and 75% 1`] = `
<div
  id="tally"
>
  <h2
    class="title"
  >
    
    electionResults
  
  </h2>
   
  <div>
    <hr
      class="hr"
    />
     
    <h3
      class="title"
    >
      mockRaceName
    </h3>
     
    <div>
      <div
        class="columns"
      >
        <div
          class="name column is-one-quarter"
        >
          
          mockName0
          
          <div
            class="party"
          >
            mockParty0
          </div>
        </div>
         
        <div
          class="votes column"
        >
          <electionresultstallybar-stub
            count="1"
            isfirst="true"
            percentage="25.0%"
          />
        </div>
      </div>
      <div
        class="columns"
      >
        <div
          class="name column is-one-quarter"
        >
          
          mockName1
          
          <div
            class="party"
          >
            mockParty1
          </div>
        </div>
         
        <div
          class="votes column"
        >
          <electionresultstallybar-stub
            count="3"
            percentage="75.0%"
          />
        </div>
      </div>
    </div>
  </div>
   
  <button
    class="button is-success tooltip"
    data-tooltip="featureNotAvailable"
  >
    downloadElectionTally
  </button>
</div>
`;
